### PR TITLE
Automatic event tracking

### PR DESCRIPTION
**Pull request by jihiggins**
An implementation of the [Automatic Event Tracking RFC](https://github.com/bevyengine/rfcs/pull/17)

Works, but if Commands are fixed to work correctly when embedded in a SystemParam, it would probably be better to remove the RwLocks in favor of using that. (Should be an easy change.)

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Ooh, our first proto-RFC! For those digging into it, this work relates closely to that done for reliable change detection (#68, #1471).

**Comment by alice-i-cecile** [comment]
We still want to add the `defer_event` functionality to this right?

**Comment by jihiggins** [comment]
Some ideas from discord:

> WrongShoe:
could you reuse the counter used for change detection to help with tracking events?

> alice:
For when your system receives an event and isn't ready to handle it yet: you want a way to put it aside without blocking all of the other later events that did get handled properly
It feels like you'd cache a clone of the extra events in the Local state, and then always check both?

(EventReader::defer_event?)

File: crates/bevy_app/src/event.rs [hunk 1]
```diff
@@ -146,10 +139,103 @@ fn map_instance_event<T>(event_instance: &EventInstance<T>) -> &T {
...
+        let mut subscriptions = world.get_resource_mut::<EventSubscriptions<T>>().unwrap();
+        subscriptions.set_subscriber_count(self.subscriber_id, self.new_count);
+    }
+}
+
 /// Reads events of type `T` in order and tracks which events have already been read.
 #[derive(SystemParam)]
 pub struct EventReader<'a, T: Component> {
-    last_event_count: Local<'a, (usize, PhantomData<T>)>,
+    commands: Commands<'a>,
```
Comment (jihiggins) [inline crates/bevy_app/src/event.rs:236]: Unclear if this works or not, will need to check outside of unit tests probably

**Comment by JeanMertz** [comment]
Two drawbacks/questions that seem worth discussing/noting:

1. Is there a performance implication to this change? Should both the old and new way of doing things be available if the new one is safer in the defined use-cases, but comes at a performance penalty?
2. What happens to the buffer if I have a reader that never reads (user error) or reads extremely infrequently?

**Comment by jihiggins** [comment]
I don't think it should impact performance very much, except for in your second point. There's a small cost when systems are added, and the current setup with the RwLock might cause some small decrease in performance. I'm not sure how much reads cost? The command setup might cost something since it locks that buffer once for each of the applied commands, but that can be fixed up later. It could also be changed to use unsafe code instead of the RwLock if it turns out to be a problem, since the reads / writes are always done w/ separate elements aside from when the readers are initially added.

For #2, it'd empty the events less often, but I think performance issues from user error is better than unexpected bugs, and it'd be relatively easy to fix (just split up the events or something.) We could add a debug assert / warning for events that have been held onto for more than x frames, if it's something we're worried about.

**Comment by jihiggins** [comment]
Re: old way / new way, I think the old way would ideally be removed if this is merged

File: crates/bevy_app/src/event.rs (hunk 1 above)
Comment (jihiggins) [reply crates/bevy_app/src/event.rs:236]: does not work 😢

**Comment by jihiggins** [comment]
The downside to this alt ManualEventReader implementation is that you have to drop it if you send events after obtaining one, since it has a reference to the Events struct. imo this isn't a huge deal and doesn't actually seem to come up very often

**Comment by jihiggins** [comment]
Event deferral is trickier than it sounded, especially if ManualEventReader stays as implemented. ManualEventReader is transient, so there's not a great way to store the deferred events. Need to think about where to cache them / how to cache them, and whether or not it's simpler if the user does it manually (you'd basically just need a Local<Vec<EventType>>.)

Added an example of manual deferral in a test, maybe could just work off of that pattern?

**Comment by alice-i-cecile** [comment]
This is compelling, but deserves a cohesive design review. Once #1662 is concluded, we should open an RFC for more reliable events.

**Comment by mockersf** [comment]
closing this PR for now as it has drifted, please reopen once the RFC is getting ready

**Pull request closed without merging**